File: 0138-copy-list-with-random-pointer/0138-copy-list-with-random-pointer.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* next;
    Node* random;
    
    Node(int _val) {
        val = _val;
        next = NULL;
        random = NULL;
    }
};
*/

class Solution {
public:
    Node* copyRandomList(Node* head) {
        if(head == NULL) return head;
        
        // modify the current linked list
        // insert new element with same value next to the old element with that value
        Node* curr = head;
        
        while(curr != NULL) {
            // insert new element 
            Node* tempNew = new Node(curr -> val);
            Node* tempNext = curr -> next;
            
            curr -> next = tempNew;
            tempNew -> next = tempNext;
            
            // skip the new element and move to next old one
            curr = tempNew -> next;
        }
        
        // now the list will contain for example 1, 1, 7, 7, 13, 13 ....
        
        // now copy the random pointers 
        curr = head;
        while(curr != NULL && curr -> next != NULL) {
            // copy random pointer of old element to random pointer of new element 
            Node* tempRand = curr -> random;
            
            // next of this random pinter will be the one to copy in new pointer
            Node* tempNew = curr -> next;
            if(tempRand != NULL) {
                tempNew -> random = tempRand -> next;
            } else {
                tempNew -> random = NULL;
            }
            
            // skip the new element and move to old one
            curr = tempNew -> next;
        }
        
        // we have copied random pointer successfully
        // now create a answer linked list 
        Node* dummyHead = new Node(-1);
        Node* tempCurr = dummyHead;
        
        curr = head;
        
        while(curr != NULL && curr -> next != NULL) {
            Node* tempNew = curr -> next;
            
            // remove new element form this list 
            curr -> next = tempNew -> next;
            
            tempCurr -> next = tempNew;
            
            tempCurr = tempCurr -> next;
            curr = curr -> next;
            
        }
        
        return dummyHead -> next;
    }
};

// tc -> O(n) + O(n) + O(n)
// sc -> O(1)

File: 0138-copy-list-with-random-pointer/0138-copy-list-with-random-pointer.go
/**
 * Definition for a Node.
 * type Node struct {
 *     Val int
 *     Next *Node
 *     Random *Node
 * }
 */

func copyRandomList(head *Node) *Node {
    if head == nil {
        return head
    }
    
    c1 := head
    var t1 *Node
    
    // first pass: add new nodes next to the old nodes
    for c1 != nil {        
        t1 = c1.Next
        c1.Next = &Node{Val: c1.Val}
        c1.Next.Next = t1
        c1 = t1
    }
    
    c1 = head
    c2 := head.Next
    // second pass: set the random pointer correctly
    for c1 != nil {
        if c1.Random == nil {
            c2.Random = nil
        } else {
            c2.Random = c1.Random.Next
        }
        
        c1 = c2.Next
        if c1 != nil {
            c2 = c1.Next
        }
    }
    
    // third pass: separate the old nodes from new
    c1 = head
    c2 = head.Next
    res := head.Next
    
    for c1 != nil { 
        c1.Next = c2.Next
        if c2.Next == nil {
            c2.Next = nil
        } else {
            c2.Next = c2.Next.Next
        }
        
        c1 = c1.Next
        c2 = c2.Next
    }
    
    return res
    
}